### PR TITLE
[Network] `az network lb address-pool` : Support subnet when creating or updating an IP-based backend pool of a load balancer.

### PR DESCRIPTION
**Pull request by msyyc**
**Description**

We have added subnet ID and IP address as a way to specify backend address when creating or updating an IP-based backend pool of a load balancer. This is related to this feature: #12937 however, we will start allowing customers to specify by subnet ID in place of vnet ID.

link issue : https://github.com/Azure/azure-cli/issues/16808
**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 1]
```diff
@@ -3337,12 +3337,14 @@
 short-summary: Create an address pool.
 parameters:
   - name: --backend-address
-    short-summary: Backend addresses information for backend address pool. If it's used, --vnet is also required.
+    short-summary: Backend addresses information for backend address pool. If it's used, --vnet is required or subnet is required.
     long-summary: |
-        Usage: --backend-address name=addr1 ip-address=10.0.0.1 --vnet MyVnet
+        Usage1: --backend-address name=addr1 ip-address=10.0.0.1 --vnet MyVnet
+        Usage2: --backend-address name=addr1 ip-address=10.0.0.1 subnet==/subscriptions/000/resourceGroups/MyRg/providers/Microsoft.Network/virtualNetworks/vnet/subnets/subnet1
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:3343]: 'Subnet' is the replace of 'vnet', so they should be in the same level. I think you should use '--subnet'.

Comment (kairu-ms) [reply src/azure-cli/azure/cli/command_modules/network/_help.py:3343]: BTW, `subnet` is belong to a `vnet`. So if a user provide a vnet name and a subnet name then you can generate the subnet ID. So `--vnet` and `--subnet` should provide both in the case.

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 2]
```diff
@@ -3509,12 +3512,22 @@ def _process_vnet_name_and_id(vnet):
...
+                                                        ip_address=addr['ip_address']) for addr in addresses_pool] if addresses_pool else None
+        except KeyError:
+            raise CLIError('Each backend address must have name, ip-address, (vnet | subnet) information.')
+    else:
+        try:
+            new_addresses = [LoadBalancerBackendAddress(name=addr['name'],
+                                                        virtual_network=VirtualNetwork(id=_process_vnet_name_and_id(addr['virtual_network'])),
+                                                        ip_address=addr['ip_address']) for addr in addresses_pool] if addresses_pool else None
+        except KeyError:
+            raise CLIError('Each backend address must have name, vnet and ip-address information.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:3529]: CLIError is not a clear error type.

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/_actions.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 import argparse
 from knack.util import CLIError
 from azure.cli.core.azclierror import UnrecognizedArgumentError
+from msrestazure.tools import is_valid_resource_id
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/network/_actions.py:11]: use `from azure.mgmt.core.tools import is_valid_resource_id` instead as we'll deprecate `msrestazure` in the future.

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 4]
```diff
@@ -3465,17 +3466,25 @@ def set_lb_frontend_ip_configuration(
...
+            resource_group=resource_group_name,
+            namespace='Microsoft.Network',
+            type='virtualNetworks',
+            name=vnet)
+    return vnet
+
+
+def _process_subnet_name_and_id(subnet, vnet, cmd, resource_group_name):
+    if subnet and not is_valid_resource_id(subnet):
+        subnet = _process_vnet_name_and_id(vnet, cmd, resource_group_name) + f'/subnets/{subnet}'
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:3482]: what if _process_vnet_name_and_id return None?

Comment (msyyc) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:3482]: There is check before here.

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**